syntax = "proto3";

import "google/api/annotations.proto";

option go_package = "github.com/sekirocc/face-recognition-service/face_service";
// option java_package = "com.sekirocc.face_recognition_service";

package com.sekirocc.face_service;

enum ResultCode {
    OK = 0;
    reserved 1 to 10000;

    SERVER_ERROR = 10001;
    reserved 10002 to 10100;

    CLIENT_ERROR = 10101;
    FACE_NOT_FOUND = 10102;
    reserved 10103 to 10200;
}

message Rect {
    Point point = 1;
    Size size = 2;
}

message Point {
    int32 x = 1;
    int32 y = 2;
}

message Size {
    int32 width = 1;
    int32 height = 2;
}

enum ImageFormat {
    IMAGE_UNKNOWN = 0;
    IMAGE_JPEG = 1;
    IMAGE_PNG = 2;
}

message Image {
    ImageFormat format = 1;
    bytes data = 2;
}

message FaceFeature {
    int32 version = 1;
    string model = 2;
    bytes blob = 3;
}

message FaceRectangle {
    Rect rectangle = 2;
    float quality = 3;
    float confidence = 4;
}

message DetectionRequest { Image image = 1; }

message DetectionResponse {
    ResultCode code = 1;
    repeated FaceRectangle face_rects = 2;
}

message ExtractionRequest { Image image = 1; }

message ExtractionResponse {
    ResultCode code = 1;
    repeated FaceRectangle face_rects = 2;
    repeated FaceFeature face_features = 3;
}

// face service
service FaceService {

    rpc Detect(DetectionRequest) returns (DetectionResponse) {
        option (google.api.http) = {
            post : "/detect_face"
            body : "*"
        };
    }

    rpc ExtractFeature(ExtractionRequest) returns (ExtractionResponse) {
        option (google.api.http) = {
            post : "/extract_face_feature"
            body : "*"
        };
    }
}
